lib/rules/no-duplicate-imports.js

```diff
@@ -54,21 +134,55 @@ function checkAndReport(context, node, value, array, messageId) {
  * @param {boolean} includeExports Whether or not to check for exports in addition to imports.
  * @param {string[]} importsInFile The array containing other imports in the file.
  * @param {string[]} exportsInFile The array containing other exports in the file.
+ * @param {{}} modulesWithImportTypes The object containing the name of unique modules with their first import type [specificImport, nameSpaceImport].
+ * @param {string[]} ExportAllDeclarationsInFile The array containing ExportAllDeclarations in the file.
  *
  * @returns {nodeCallback} A function passed to ESLint to handle the statement.
  */
-function handleImports(context, includeExports, importsInFile, exportsInFile) {
+function handleImports(
+    context,
+    includeExports,
+    importsInFile,
+    exportsInFile,
+    modulesWithImportTypes,
+    ExportAllDeclarationsInFile
+) {
     return function(node) {
         const value = getValue(node);
+        const type = getImportType(node);
 
         if (value) {
-            checkAndReport(context, node, value, importsInFile, "import");
+            checkAndReport(
+                context,
+                node,
+                value,
+                importsInFile,
+                "import",
+                modulesWithImportTypes,
+                type,
+                ExportAllDeclarationsInFile
+            );
 
             if (includeExports) {
-                checkAndReport(context, node, value, exportsInFile, "importAs");
+                checkAndReport(
+                    context,
+                    node,
+                    value,
+                    exportsInFile,
+                    "importAs",
+                    modulesWithImportTypes,
+                    type,
+                    ExportAllDeclarationsInFile
+                );
             }
-
             importsInFile.push(value);
+            if (modulesWithImportTypes[value]) {
+                modulesWithImportTypes[value] = modulesWithImportTypes[
+                    value
+                ].concat(type);
+            } else {
+                modulesWithImportTypes[value] = [type];
+            }
```

Comment:
Do we need both `importsInFile` and `modulesWithImportTypes` now, it seems redundant. Can we instead maintain only one map? The key would be the module source, the value would be an array of import declarations with that path.

Reply:
Yes, it’s redundant, we can use one variable in this case our map with source module and declarations, then we will remove these variables: **importsInFile**, **modulesWithImportTypes**, **exportsInFile** and **exportAllDeclarationsInFile**.